archive/miu_language/decision_nec.lean

```diff
@@ -0,0 +1,310 @@
+-/
+def goodm (xs : miustr) : Prop :=
+  ∃ ys : miustr, xs = (M::ys) ∧ ¬(M ∈ ys)
+
+/- Example usage -/
+
+lemma goodmi : goodm [M,I] :=
+begin
+  split,
+    swap,
+  exact [I],
+  simp,
+end
+
+/-
+We'll show, for each `i` from 1 to 4, that if `en` follows by rule `i` from `st` and if
+`goodm st` holds, then so does `goodm en`.
+-/
+
+
+open list
+
+lemma goodmrule1 (xs : miustr) (h₁ : derivable (xs ++ [I])) (h₂ : goodm (xs ++ [I]))
+  : goodm (xs ++ [I,U]) :=
+begin
+  rcases h₂ with ⟨ys, k₁, k₂⟩,
+  use ys ++ [U],
+  split,
+    rw [←(cons_append _ _ _), ←k₁], simp,
+  simp [k₂],
+end
+
+
+lemma goodmrule2 (xs : miustr) (h₁ : derivable (M :: xs))
+  (h₂ : goodm (M :: xs)) : goodm (M :: xs ++ xs) :=
+begin
+  rcases h₂ with ⟨ys, k₁, k₂⟩,
+  use (ys ++ ys),
+  split,
+    rw (cons_inj _).mp k₁,
+    refl,
+  simp [k₂],
+end
+
+
+lemma goodmrule3  (as bs : miustr) (h₁ : derivable (as ++ [I,I,I] ++ bs))
+  (h₂ : goodm (as ++ [I,I,I] ++ bs)) : goodm (as ++ U :: bs) :=
+begin
+  rcases h₂ with ⟨ys, p₁, p₂⟩,
+  change as ++ [I,I,I] ++ bs with (as ++ [I,I,I]) ++ bs at p₁,
+  have h : ∃ zs, as = M :: zs,
+    induction as with x xs hxs, { -- base case
+      exfalso,
+      have : I = M,
+        exact head_eq_of_cons_eq p₁,
+      contradiction,
+    }, {
+      use xs,
+      rw head_eq_of_cons_eq p₁,
+    },
+  cases h with zs h,
+  use (zs++[U]++bs),
+  split, {
+    simp [h],
+  }, {
+    rw h at p₁,
+    rw ←((cons_inj M).mp p₁) at p₂,
+    simp [append_eq_has_append] at p₂,
+    simp [mem_append] at *,
+    exact p₂,
+  },
+end
+
+
+-- The proof of the next lemma is very similar to the previous proof!
+
+lemma goodmrule4  (as bs : miustr) (h₁ : derivable (as ++ [U,U] ++ bs))
+  (h₂ : goodm (as ++ [U,U] ++ bs)) : goodm (as ++ bs) :=
+begin
+  rcases h₂ with ⟨ys, p₁, p₂⟩,
+  change as ++ [U,U] ++ bs with (as ++ [U,U]) ++ bs at p₁,
+  have h : ∃ zs, as = M :: zs,
+    induction as with x xs hxs, { -- base case
+      exfalso,
+      have : U = M,
+        exact head_eq_of_cons_eq p₁,
+      contradiction,
+    }, {
+      use xs,
+      rw head_eq_of_cons_eq p₁,
+    },
+  cases h with zs h,
+  use (zs++bs),
+  split, {
+    rw h, refl,
+  }, {
+    rw h at p₁,
+    rw ←((cons_inj M).mp p₁) at p₂,
+    simp [append_eq_has_append] at p₂,
+    simp [mem_append] at *,
+    exact p₂,
+  }
+end
+
+
+/--
+Any derivable string must begin with `M` and contain no `M` in its tail.
+-/
+
+
+/-
+z : miustr,
+k : ℕ,
+hk : derivable (z ++ repeat U (k * 2)) → derivable z,
+h : derivable (z ++ (repeat U (k * 2) ++ [U, U]))
+⊢ derivable (z ++ repeat U (k * 2))
+-/
+
+/-
+z : miustr,
+k : ℕ,
+hk : derivable (z ++ repeat U (k * 2)) → derivable z,
+h : derivable (z ++ (repeat U (k * 2) ++ [U, U]))
+⊢ derivable (z ++ repeat U (k * 2))
+-/
+
+/-
+z : miustr,
+k : ℕ,
+hk : derivable (z ++ repeat U (k * 2)) → derivable z,
+h : derivable (z ++ (repeat U (k * 2) ++ [U, U]))
+⊢ derivable (z ++ repeat U (k * 2))
+-/
```

Comment:
What is this?

Reply:
Woops. I must have hit the 'copy state to comment' button. Will remove.